libgnucash/engine/gnc-features.c

```diff
@@ -48,6 +48,7 @@ static gncFeature known_features[] =
     { GNC_FEATURE_GUID_BAYESIAN, "Use account GUID as key for Bayesian data (requires at least GnuCash 2.6.12)" },
     { GNC_FEATURE_GUID_FLAT_BAYESIAN, "Use account GUID as key for bayesian data and store KVP flat (requires at least Gnucash 2.6.19)" },
     { GNC_FEATURE_SQLITE3_ISO_DATES, "Use ISO formatted date-time strings in SQLite3 databases (requires at least GnuCash 2.6.20)"},
+    { GNC_FEATURE_REG_SORT_FILTER, "Store the Register sort and filter settings in .gcm file only (requires at least GnuCash 4.0)"},
```

Comment:
Not sure about this, with these changes the sort and filter are saved to both .gcm and kvp entries with loading from .gcm and then kvp if no entry found. So in master leading to 4.0 I would set this feature to true when a conversion routine ran moving entries from kvp to .gcm and a change to only saving and loading from .gcm file.

Reply:
So all I need to do for this PR is amend the feature text and push that along with the other changes. Then wait for 3.3 and do the conversion, delete old keys and set the feature flag.

Reply:
No. Gnucash 3.x should **never** set the feature flag by itself because that would mean older early 3.x version and 2.6.21 won't be able to work on such a book any more. We try to maintain this backwards compatibility to the last release of the previous stable series.

But gnucash 3.3 and up should behave differently if it's set.

It's a difficult matter to do right. And while thinking this through again I believe the behaviour should still be slightly different than what I suggested.
Perhaps the easiest way to visualize it is that the feature flag's presence defines the storage format of the filter/sort stuff. If the flag is absent, filter/sort should go in kvp. If the flag is set, filter/sort should go in .gcm. No version of gnucash should ever write to both formats at the same time.

So what format 3.x will use depends on the feature flag being present or not. If the flag is set, 3.3 and up should only interact with .gcm. If the flag is not set, 3.x should only interact with kvp (which works for the complete 3.x series and even 2.6.21).

So 3.x should be able to read and write both kvp and .gcm, but never at the same time. The feature flag defines which format to use.

The conversion from kvp to .gcm can only happen in a 4.x release, the first time a book without feature flag set is opened. Again no 3.x release should ever do the conversion.

That should allow you to cleanly separate the two implementations. In 3.3 you add both implementations and select one or the other depending on the feature flag (you may have to extract saving and loading of filters to separate functions to make this easier, I didn't recheck).

In 4.0 you only use the kvp functions in a one-time migration function. This will
1. to read the old values from kvp
1. clear the kvp
1. write the values to .gcm
1. set the feature flag

Reply:
Will sleep on this....